### PR TITLE
Require matching data types in MDArray::read_into_slice

### LINKED ISSUE #290: MDArray: Different read type compared to array type leads to ASAN issues
From https://github.com/georust/gdal/pull/284:

@lnicola:
> Try this under ASAN:
>
>             let string_array = root_group
>                .open_md_array("uint_var", CslStringList::new())
>                .unwrap();
>
>             let values = string_array.read_as::<u8>(vec![0, 0], vec![1, 2]).unwrap();
>
> This happens because GDALMDArrayRead takes the buffer type as argument, but we're passing GDALMDArrayGetDataType(self.c_mdarray), which is the array type. Normally, GDAL does some format conversions when needed. When we call GDALRasterIOEx, we pass T::gdal_type() instead.

Maybe we can also make sure that the read type aligns with the actual requested and returned type.
However, we should keep in mind that we can get strings and complex types as return values.

### PR DESCRIPTION
**Pull request by lnicola**
- [x] I agree to follow the project's [code of conduct](https://github.com/georust/gdal/blob/master/CODE_OF_CONDUCT.md).
- [x] I added an entry to `CHANGES.md` if knowledge of this change could be valuable to users.
---

We don't support `GEDTC_COMPOUND` anyway because of the `GdalType` bound. ~~This is stricter than it could be, but should be fine as a quick workaround for #290. We should pass in the data type of the buffer, but it wasn't obvious to me how to do that.~~

Fixes #290

### CODE REVIEW

File: src/raster/mdarray.rs [hunk 1]
```diff
@@ -162,6 +162,16 @@ impl<'a> MDArray<'a> {
...
 
+            // HACK: this is too strict, but see https://github.com/georust/gdal/issues/290
+            if !self.datatype().class().is_numeric()
+                || self.datatype().numeric_datatype() != T::gdal_type()
+            {
+                return Err(GdalError::UnsupportedMdDataType {
+                    data_type: self.datatype().class(),
+                    method_name: "GDALMDArrayRead",
+                });
+            }
```
Comment (ChristianBeilschmidt) [inline src/raster/mdarray.rs:173]: Maybe we can have the call of `datatype` only once.
Now, you have one C call and two fn-calls that wrap the C call.
(we could move this out of the unsafe block, then)

Moreover, can you elaborate why this is too strict?

**Review by ChristianBeilschmidt — CHANGES_REQUESTED** [review]

File: src/raster/mdarray.rs (hunk 1 above)
Comment (lnicola) [reply src/raster/mdarray.rs:173]: It's too strict if it works like the legacy API. There you could read an `u16` band into an `u8` array, or the other way around, and GDAL would do a reasonable conversion (by saturating the values). This implementation prevents that.

The solution would be to pass the buffer data type, but I don't know how to get a `GDALExtendedDataTypeH` from a `GDALDataType`.

**Review by ChristianBeilschmidt — APPROVED** [review]

**Comment by ChristianBeilschmidt** [comment]
Looks more robust now.
If this resolves the ASAN issue https://github.com/georust/gdal/issues/290, I would like to merge it.

**Comment by jdroenner** [comment]
bors merge

**Comment by lnicola** [comment]
> If this resolves the ASAN issue https://github.com/georust/gdal/issues/290

It does, I included that part in `test_read_string_array`.

It actually didn't need ASAN to crash, it was a blatant buffer overrun.

**Pull request merged**